### PR TITLE
DrivAerML: T_max=5 + gc=1.0 + WD=1e-2 — AirfRANS golden cycling on 3D

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AirfRANS uses T_max=5 cosine cycling in its golden config (0.001236 baseline). DrivAerML uses T_max=30. The AirfRANS recipe transfer theme tests whether AirfRANS ingredients work on DrivAerML. T_max=5 is the fastest cycling tested on AirfRANS and produces the most annealing cycles per training budget. On DrivAerML with 394 batches/epoch (vs ~360 for AirfRANS), T_max=5 would complete 1 full cosine cycle every 5 epochs — rapid exploration of the loss landscape. Combined with gc=1.0 and WD=1e-2 from the AirfRANS golden recipe, this tests the complete scheduler transfer.

## Instructions

**IMPORTANT: Use EXACTLY these run specifications.**

Run 1 — DrivAerML T_max=5 + gc=1.0 + WD=1e-2 (full AirfRANS scheduler transfer):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 5 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group zenitsu-tmax5-drivaer
```

Run 2 — DrivAerML T_max=5 + gc=1.0 only (no WD, isolate scheduler effect):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 5 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group zenitsu-tmax5-drivaer
```

Run 3 — DrivAerML T_max=7 + gc=1.0 + WD=1e-2 (intermediate between 5 and 10):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 7 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group zenitsu-tmax5-drivaer
```

Run 4 — AirfRANS T_max=5 + gc=1.0 + WD=1e-2 (verify AirfRANS baseline with 2L/256d):
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 5 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --seed 43 \
  --wandb-group zenitsu-tmax5-drivaer
```

Report `val_primary/surface_rel_l2_pct` for DrivAerML runs and `val_primary/surface_mse` for AirfRANS.

## Baseline

- **DrivAerML val_primary/surface_rel_l2_pct:** 4.619% (PR #2691, epoch 256)
- **AirfRANS val_primary/surface_mse:** 0.001236 (PR #2828, epoch 349)
- **DrivAerML external target:** 3.71% (AB-UPT)
- **Reproduce golden DrivAerML:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT zenitsu:

## Results

### DrivAerML — `val_primary/surface_rel_l2_pct` (baseline: 4.619%, external target: 3.71% AB-UPT)

| Run | Config | Best | Best Epoch | Diverged At | W&B ID |
|-----|--------|------|------------|-------------|--------|
| 1 | T_max=5 + gc=1.0 + **WD=1e-2** | 7.02% | 132 | Epoch 133 (NaN by 137) | `n69ye17p` |
| 2 | T_max=5 + gc=1.0 only (**no WD**) | **5.38%** | 180 | Epoch 198 (grad norm 765→14K→39K) | `5jyc79dr` |
| 3 | T_max=7 + gc=1.0 + WD=1e-2 | 9.08% | 90 | Epoch 93 (NaN by 100) | `ywuqzxom` |

### AirfRANS — `val_primary/surface_mse` (baseline: 0.001236)

| Run | Config | Best | Best Epoch | Status | W&B ID |
|-----|--------|------|------------|--------|--------|
| 4 | T_max=5 + gc=1.0 + WD=1e-2 (2L/256d) | **0.001976** | 244 | Stable at epoch 320+ | `r028up1n` |

### Peak Memory

- DrivAerML runs (GPUs 0–2): ~6.0 GB each
- AirfRANS run (GPU 3): ~22.7 GB

### Commands Used

Exact commands from PR instructions with `--agent zenitsu` and `--wandb-name` added. Runs launched in parallel on GPUs 0–3 via `CUDA_VISIBLE_DEVICES`.

**Run 1 (GPU 0):**
```
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 5 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --weight-decay 1e-2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent zenitsu --wandb-name "zenitsu/tmax5-gc1-wd1e2-drivaer" --wandb-group zenitsu-tmax5-drivaer
```

**Run 2 (GPU 1):** Same as Run 1 but without `--weight-decay 1e-2`, wandb-name `"zenitsu/tmax5-gc1-no-wd-drivaer"`

**Run 3 (GPU 2):** Same as Run 1 but `--cosine-t-max 7`, wandb-name `"zenitsu/tmax7-gc1-wd1e2-drivaer"`

**Run 4 (GPU 3):**
```
CUDA_VISIBLE_DEVICES=3 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --seed 43 --agent zenitsu --wandb-name "zenitsu/airfrans-tmax5-gc1-wd1e2" --wandb-group zenitsu-tmax5-drivaer
```

### What Happened

**The AirfRANS golden cycling recipe (T_max=5) does not transfer to DrivAerML.** All three DrivAerML configurations eventually diverged catastrophically via exploding gradients.

**Divergence pattern:** Each run showed a consistent progression — gradually improving metrics with each cosine cycle, then a sudden gradient explosion (grad norm jumping from ~0.3 to hundreds/thousands over 2–3 epochs), followed by NaN within a few more epochs. The divergence appears to be triggered by the sharp LR restart at the start of a new cosine cycle, where the model has overfit to the current cycle's minimum and the sudden LR jump destabilizes it.

**Key findings:**

1. **T_max=5 cycling is fundamentally unstable on DrivAerML long-term.** All 3 DrivAerML runs diverged. The rapid cycling (every 5 epochs) may cause excessive parameter oscillation that accumulates over many cycles.

2. **Weight decay (1e-2) worsened stability.** Run 2 (no WD) survived until epoch 198 and achieved 5.38%. Run 1 (with WD) diverged 65 epochs earlier at epoch 133, reaching only 7.02%. The WD=1e-2 value from AirfRANS appears too aggressive for DrivAerML's 4L/512d model.

3. **T_max=7 was least stable** — diverged earliest at epoch 93 with best of only 9.08%.

4. **Before divergence, the runs showed promising improvement** — Run 2 reached 5.38% at epoch 180, improving from ~11% at epoch 50 to ~5.4% by epoch 180. The cycling mechanism does drive improvement, but instability eventually kills the run.

5. **AirfRANS is stable with the same recipe** — Run 4 (T_max=5+gc+WD on AirfRANS 2L/256d) is healthy at epoch 320+ with no instability, reaching 0.001976 (vs baseline 0.001236, so 60% worse). The instability appears specific to DrivAerML's larger model or data characteristics.

6. **DrivAerML baseline (T_max=30) is superior.** The baseline achieved 4.619% at epoch 256 with no divergence. T_max=30 gives longer cosine periods that avoid the instability from rapid cycling.

### Suggested Follow-ups

1. **Try T_max=10 or T_max=15** — intermediate between the unstable T_max=5/7 and the stable baseline T_max=30. This could find a sweet spot that captures cycling benefits without divergence.
2. **Lower gradient clip threshold** — gc=0.5 or gc=0.3 might prevent the divergence cascade while preserving the cycling benefits. The divergence always started with grad norms climbing from 0.3 → 0.5 → 1.0+ before exploding.
3. **Cosine annealing with warm restarts (SGDR-style)** — instead of hard restarts, use a gradual warm-up at the start of each cycle to avoid the destabilizing LR spike.
4. **Lower LR for faster cycling** — the combination of high LR (5e-4) and fast cycling (T_max=5) may be too aggressive. Try LR=2e-4 with T_max=5.
5. **AirfRANS seed sensitivity** — Run 4 with seed 43 reached 0.001976 vs baseline 0.001236 (seed 42). The difference could be seed sensitivity rather than a recipe issue — worth verifying with the original seed.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2911 because ADVISOR: Closing — T_max=5 is definitively too aggressive for DrivAerML 4L/512d:

| Run | Config | Best | Fate |
|---|---|---|---|
| Run 1 | T_max=5+gc=1.0+WD=1e-2 | 7.02% | NaN ep137 |
| Run 2 | T_max=5+gc=1.0 (no WD) | 5.38% | diverged ep230 |
| Run 3 | T_max=7+gc=1.0+WD=1e-2 | 9.08% | NaN ep100 |
| Run 4 | AF T_max=5+gc=1.0+WD=1e-2 | 0.001976 | unstable |

All DM runs diverged via gradient explosion at cosine cycle boundaries. WD=1e-2 worsened instability (earlier crash in Run 1 vs Run 2). T_max=7 also crashed. The AirfRANS recipe's rapid cycling does NOT transfer to DM's 4L/512d scale.

Your diagnosis was correct: DM's larger model needs gentler cycling. T_max=10-15 is the right exploration range — bulma #2879 (T_max=10), chopper #2882 (T_max=15), and einar #2883 (T_max=20) are testing these. I'll assign you a fresh direction. Thank you..
